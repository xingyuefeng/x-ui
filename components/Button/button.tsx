import React, { ReactNode, FC, ButtonHTMLAttributes, AnchorHTMLAttributes, forwardRef, MutableRefObject } from 'react';
import classnames from 'classnames';

export type ButtonSize = 'lg' | 'sm'

export type ButtonType = 'primary' | 'default' | 'danger' | 'link'

interface BaseButtonProps  {
  children: ReactNode
  disabled?: boolean;
  href?: string;
  size?: ButtonSize;
  type?: ButtonType;
  ref?: MutableRefObject<HTMLButtonElement> &  MutableRefObject<HTMLAnchorElement> ;
}

type NativeButtonProps = BaseButtonProps & Omit<ButtonHTMLAttributes<HTMLElement>, 'type'>
type AnchorButtonProps = BaseButtonProps & AnchorHTMLAttributes<HTMLElement>

type ButtonProps = Partial<NativeButtonProps & AnchorButtonProps>


const Button: FC<ButtonProps> = forwardRef((props, ref) => {
  const {
    children,
    className,
    disabled,
    href,
    size,
    type,
    ...restProps
  } = props;
  const classes = classnames('btn', className, {
    [`btn-${type}`]: type,
    [`btn-${size}`]: size,
    'disabled': (type === 'link') && disabled
  })
  if (type === 'link' && href) {
    return (
    <a
      className={classes}
      href={href}
      ref={ref}
      {...restProps}
    >
      {children}
    </a>)
  } else {
    return (
    <button
      className={classes}
      disabled={disabled}
      ref={ref}
      {...restProps}
    >
      {children}
    </button>)
  }
})

Button.defaultProps = {
  type: 'default',
  disabled: false,
}


export default Button