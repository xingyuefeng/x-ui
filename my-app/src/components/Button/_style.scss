.btn {
  display: inline-block;
  font-size: $btn-font-size;
  line-height: $btn-line-height;
  font-weight: $btn-font-weight;
  border: $border-width solid transparent;
  box-shadow: $btn-box-shadow;
  cursor: pointer;
  transition: $btn-transition;
  white-space: nowrap;
  outline: none;
  @include button-size($btn-padding-y, $btn-padding-x,$font-size-base, $border-radius);
  &.disabled,
  &[disabled] {
    cursor: not-allowed;
    opacity: $btn-disabled-opacity;
    box-shadow: none;
    > * {
      pointer-events: none;
    }
  }
}

.btn-primary {
  @include button-style($primary, $white);
}

.btn-default {
  @include button-style($dark, $black, lighten($gray-500, 10%));
}

.btn-danger {
  @include button-style($danger, $white);
}

.btn-link {
  color: $btn-link-color;
  box-shadow: none;
  text-decoration: none;
  &:hover {
    color: $btn-link-hover-color;
  }
  &:focus {
    color: $btn-link-hover-color;
  }
}

.btn-lg {
  @include button-size($btn-padding-y-lg, $btn-padding-x-lg, $btn-font-size-lg, $border-radius-lg)
}

.btn-sm {
  @include button-size($btn-padding-y-sm, $btn-padding-x-sm, $btn-font-size-sm, $border-radius-sm)
}

