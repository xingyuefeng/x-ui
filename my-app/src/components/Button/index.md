# Button 按钮


## 基本用法

```jsx
import { Button } from 'zbui';

ReactDOM.render(
  <>
    <Button>default</Button>
    <Button type="primary">primary</Button>
     <Button type="danger">danger</Button>
     <Button type="link">link</Button>
  </>
, mountNode);
```

## 尺寸

```jsx
import { Button } from 'zbui';

ReactDOM.render(
  <>
    <Button size="lg">default</Button>
    <Button size="small">primary</Button>
  </>
, mountNode);
```


## API

| 属性 | 类型 | 默认值 | 说明 |
| :--- | :--- | :--- | :--- |
| type | string | 'default' | 设置主题，可选值为 `default`、`primary`、`danger`、 `link` |
| size | string | -' | 设置大小，可选值为 `lg`、`sm` |
| disabled | boolean | false | 是否禁用 |
支持原生button其他属性