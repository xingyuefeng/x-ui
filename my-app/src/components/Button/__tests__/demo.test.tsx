import React from 'react';
import { mount, render, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Button from '..'
import { ButtonSize, ButtonType }  from '../button'

describe('Button', () => {
it('render correctly', () => {
    const wrapper = render(<Button>hello</Button>);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  // it('warns if size is wrong', () => {
  //   //使用jest.fn()对方法进行mock
  //   const mockWarn = jest.fn();
  //   //使用jest.spyOn()模拟跟踪某个类的方法的调用
  //   jest.spyOn(console, 'warn').mockImplementation(mockWarn);
  //   const size = ('who am I' as any) as ButtonSize;
  //   render(<Button size={size} />);
  //   expect(mockWarn).toHaveBeenCalledTimes(1);
  //   // 第一次调用时的第一个参数是0
  //   expect(mockWarn.mock.calls[0][0]).toMatchObject({
  //     message: 'unreachable case: "who am I"',
  //   });
  // });
it('should support link button', () => {
    const wrapper = mount(
      <Button target="_blank" href="http://1.com" type="link">
        link button
      </Button>,
    );
    expect(wrapper.render()).toMatchSnapshot();
  });

it('disabled', () => {
    const wrapper = render(<Button disabled>foo</Button>);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

it('onClick', () => {
    const onClick = jest.fn();
    const wrapper = mount(<Button onClick={onClick}>foo</Button>);
    wrapper.simulate('click');
    expect(onClick).toHaveBeenCalled();
  });
it('onClick when disabled', () => {
    const onClick = jest.fn();
    const wrapper = mount(<Button disabled={true} onClick={onClick}>foo1</Button>);
    wrapper.simulate('click');
    expect(onClick).not.toHaveBeenCalled()
  })
})



