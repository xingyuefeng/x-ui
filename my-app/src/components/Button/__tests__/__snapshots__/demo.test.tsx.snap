// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button disabled 1`] = `
<button
  class="btn btn-default"
  disabled=""
>
  foo
</button>
`;

exports[`Button render correctly 1`] = `
<button
  class="btn btn-default"
>
  hello
</button>
`;

exports[`Button should support link button 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "btn btn-link",
      "href": "http://1.com",
      "target": "_blank",
    },
    "children": Array [
      Object {
        "data": "link button",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "text",
      },
    ],
    "name": "a",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
      "href": undefined,
      "target": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
      "href": undefined,
      "target": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
